internal/connection/establish_connection.go

```diff
@@ -2,19 +2,51 @@ package connection
 
 import (
 	"fmt"
-	"github.com/psiayn/heiko/internal/config"
-	"github.com/spf13/viper"
-	"golang.org/x/crypto/ssh"
+	"io/ioutil"
 	"log"
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/psiayn/heiko/internal/config"
+	"github.com/spf13/viper"
+	"golang.org/x/crypto/ssh"
 )
 
+func publicKey(path string) []ssh.AuthMethod {
+	if path == "" {
+		home, err := os.UserHomeDir()
+		if err != nil {
+			panic(err)
+		}
+		path = filepath.Join(home, ".ssh/heiko/key")
+	}
+
+	key, err := ioutil.ReadFile(path)
+	if err != nil {
+		return []ssh.AuthMethod{}
+	}
+
+	signer, err := ssh.ParsePrivateKey(key)
+	if err != nil {
+		return []ssh.AuthMethod{}
+	}
+
+	return []ssh.AuthMethod{ssh.PublicKeys(signer)}
+}
+
 func Connect(node config.Node) (*ssh.Client, error) {
+
 	sshConfig := &ssh.ClientConfig{
 		User: node.Username,
-		Auth: []ssh.AuthMethod{ssh.Password(node.Password)},
+	}
+
+	switch strings.ToUpper(node.Auth.Method) {
+	case "PASSWORD":
+		sshConfig.Auth = []ssh.AuthMethod{ssh.Password(node.Auth.Password)}
+
+	case "KEYS":
+		sshConfig.Auth = publicKey(node.Auth.Keys.Path)
```

Comment:
There's a bug here. When `auth` isn't specified in the config (i.e., how the config looked before this PR), I get an error like this in the logs
```
ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
```

Since the scheduler retries if it fails, these errors keep repeating and blow up the log. I'm guessing it's because there's no default case here, but I'm not really sure.

Reply:
The same issue also occurs when `method: "keys"` is given without a key specified and `heiko start` is run before `heiko init`. We should document this somewhere and/or provide a better error message.

Reply:
I noticed the same error in yet another case.
```
ERROR while connecting to node: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
```

When a key is specified in the config:
```yaml
      method: "keys"
      keys:
        path: "~/.ssh/id_rsa"
```
but the given key is not in the node's `authorized_keys`.